sql-plugin/src/main/delta-lake-common/scala/com/nvidia/spark/rapids/delta/RapidsDeltaWrite.scala

```diff
@@ -0,0 +1,132 @@
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+    case p: RapidsDeltaWrite =>
+      Seq(RapidsDeltaWriteExec(planLater(p.child)))
+    case _ => Nil
+  }
+}
+
+/**
+ * This logical node is a placeholder in the plan for where, semantically, a Delta Lake
+ * write occurs. Normally Delta Lake writes are *not* visible in the SQL plan, but we
+ * use this as a workaround for undesired row transitions in the plan when adaptive
+ * execution is enabled. Without it, AdaptiveSparkPlanExec can become the root node
+ * of the physical plan to write into Delta. Without a parent node above
+ * AdaptiveSparkPlanExec, the GPU transition planning does not know whether to make
+ * the result of the adaptive plan GPU columnar or CPU row, so it ends up assuming CPU
+ * row and injecting a columnar-to-row transition at the end of the plan. The Delta
+ * write code cannot directly replace the AdaptiveSparkPlanExec to force it to be
+ * columnar, because it's tied to the QueryExecution which can only be created with
+ * logical plans. Thus our workaround is to inject a node in the logical plan for
+ * the write and ensure the physical node appears to Spark as a write so the
+ * AdaptiveSparkPlanExec node is a child of it rather than the parent. This provides
+ * the necessary parent node for the AdaptiveSparkPlanExec so the GPU transition
+ * planning code can correctly determine whether the AQE plan should be GPU columnar
+ * or CPU row.
+ */
+case class RapidsDeltaWrite(child: LogicalPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+
+  override def withNewChildInternal(newChild: LogicalPlan): LogicalPlan = {
+    copy(child = newChild)
+  }
+}
+
+/**
+ * A stand-in physical node for a Delta write. This should *always* be replaced by
+ * a GpuRapidsDeltaWriteExec during GPU planning and should never be executed.
+ */
+case class RapidsDeltaWriteExec(child: SparkPlan) extends V2CommandExec with UnaryExecNode {
+  override def output: Seq[Attribute] = child.output
+
+  override def run(): Seq[InternalRow] = {
+    throw new IllegalStateException("Should have been replaced with a GPU node before execution")
+  }
+
+  override def withNewChildInternal(newChild: SparkPlan): SparkPlan = {
+    copy(child = newChild)
+  }
+}
+
+/**
+ * A stand-in physical node for a GPU Delta write. It needs to appear as a writing
+ * command or data source command so the AdaptiveSparkPlanExec does not appear at the
+ * root of the plan to write. See the description for RapidsDeltaWrite for details.
+ */
+case class GpuRapidsDeltaWriteExec(child: SparkPlan) extends V2CommandExec
+    with UnaryExecNode with GpuExec {
+  override def output: Seq[Attribute] = child.output
+
+  lazy val basicMetrics: Map[String, SQLMetric] = GpuWriteJobStatsTracker.basicMetrics
+  lazy val taskMetrics: Map[String, SQLMetric] = GpuWriteJobStatsTracker.taskMetrics
+
+  override lazy val allMetrics: Map[String, GpuMetric] =
```

Comment:
Because we are exposing actual metrics do we want to explain somewhere in the user docs what this node is and that the metrics are not really all that relevant.

Reply:
I'll add something to the Delta Lake docs, but why would the metrics be not relevant? These are the same metrics that a typical, non-Delta write shows, and they are updated via the actual write code (see the extra job tracker added in GpuOptimisticTransaction.writeFiles using the metrics from this node).

Reply:
Where are the metrics getting set?  they are not set here. This is not doing any writes, it just passes the buck on to its child to execute.

`GpuWriteJobStatsTracker.basicMetrics` is a `def` that will eventually call down and produce a map of metrics that calls `create*` to produce the `GpuMetric` instances.  That `create*` is going to produce an ID for each of the metrics. The ID is the critical part to make sure everything lines up so the info from the task ends up tied back the proper node in the SQL graph. For `number of written files` to be updated for the `GpuRapidsDeltaWriteExec` node in SQL UI, the metrics or at least the IDs from `GpuRapidsDeltaWriteExec.allMetrics` must be passed into the operator that actually does the write. If you do manage to make this work, then we should try to make the other node in the tree NOT include the ID, or else some automated tools might get really confused seeing the same metric/ID in two places in the same graph.  Best case it end up double counting some things.

Reply:
> Where are the metrics getting set?

They are getting set by the job stats tracker being added to the GpuFileFormat.write call.  See https://github.com/NVIDIA/spark-rapids/pull/7506/files#diff-fa589a4afea438f0bc720a61e640480ccd9a38a3f40050978d3957e740572e6cR246-R250

> The ID is the critical part to make sure everything lines up so the info from the task ends up tied back the proper node in the SQL graph.

Yes, I'm relying on this to tie the metrics being tracked by the added job stats tracker in the GpuFileFormat.write call back to this node in the SQL UI.

> For number of written files to be updated for the GpuRapidsDeltaWriteExec node in SQL UI, the metrics or at least the IDs from GpuRapidsDeltaWriteExec.allMetrics must be passed into the operator that actually does the write.

That's what's happening in the code I linked to above.

> If you do manage to make this work, then we should try to make the other node in the tree NOT include the ID,

What other node?  Delta write normally does not appear at all in a SQL plan, it's hidden as part of some larger operation, like a SaveIntoDataSourceCommand or MergeIntoDelta, that appears in a completely different SQL plan that executes this plan as a sub-plan.  Those outer-plan nodes have their own separate metrics with their own separate IDs.  That's why I added a whole new stats tracker to the write instead of overwriting/reusing an existing one.

Reply:
Okay so long as you have tested this and are sure that it is working.